clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -191,15 +195,22 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
                     } else {
                         writeOriginalBatch = false;
                     }
+
+                    if (!record.hasValue()) {
+                        containsTombstones = true;
+                    }
                 }
             }
 
             if (!retainedRecords.isEmpty()) {
-                if (writeOriginalBatch) {
+                // we check if the delete horizon should be set to a new value
+                // in which case, we need to reset the base timestamp and overwrite the timestamp deltas
+                // if the batch does not contain tombstones, then we don't need to overwrite batch
+                if (writeOriginalBatch && (!shouldSetDeleteHorizon || !containsTombstones)) {
                     batch.writeTo(bufferOutputStream);
                     filterResult.updateRetainedBatchMetadata(batch, retainedRecords.size(), false);
                 } else {
-                    MemoryRecordsBuilder builder = buildRetainedRecordsInto(batch, retainedRecords, bufferOutputStream);
+                    MemoryRecordsBuilder builder = buildRetainedRecordsInto(batch, retainedRecords, bufferOutputStream, firstTimestamp);
```

Comment:
We probably only want to set the DeleteHorizonTime if the batch contains tombstone.

Reply:
Yeah, I missed that. Should've also checked in the else part of the code as well.